File: pages/20260824170700_holding_topology_passing/analysis/julia/passing_occupancy.jl
function split_interval(start, duration, period)
    duration == 0 && return Tuple{Rational{Int},Rational{Int}}[]
    duration >= period && return [(0 // 1, period)]
    normalized = mod(start, period)
    finish = normalized + duration
    finish <= period && return [(normalized, finish)]
    return [(normalized, period), (0 // 1, finish - period)]
end

function occupancy_shares(schedule::PassingSchedule; dwell_ratio=DWELL_RATIO)
    period = schedule.cycle_length // 1
    dwell = schedule.hand_period * dwell_ratio
    intervals = Dict{Int,Vector{Tuple{Rational{Int},Rational{Int}}}}()
    for event in schedule.cycle_tosses
        event.kind == "empty" && continue
        event.ball === nothing && continue
        object_intervals = get!(intervals, event.ball, Tuple{Rational{Int},Rational{Int}}[])
        if event.kind == "hold"
            append!(object_intervals, split_interval(event.beat // 1, event.height // 1, period))
        else
            append!(object_intervals, split_interval(event.beat + event.height - dwell, dwell, period))
        end
    end
    objects = sort(collect(keys(intervals)))
    segments = [intervals[object] for object in objects]
    boundaries = Rational{Int}[0 // 1, period]
    for object_segments in segments, (start, finish) in object_segments
        push!(boundaries, start, finish)
    end
    sort!(unique!(boundaries))
    occupancy_ticks = zeros(Rational{Int}, schedule.ball_count + 1)
    for (start, finish) in zip(boundaries[1:end-1], boundaries[2:end])
        held = count(object_segments -> any(
            held_start <= start && finish <= held_finish
            for (held_start, held_finish) in object_segments
        ), segments)
    occupancy_ticks[held + 1] += finish - start
    end
    shares = occupancy_ticks ./ period
    return (
        object_count=schedule.ball_count,
        occupancy_shares=shares,
        p_alpha=first(shares),
        p_amphoteron=sum(shares[2:end-1]; init=0 // 1),
        p_kappa=last(shares),
    )
end

function body_occupancy_shares(schedule::PassingSchedule; dwell_ratio=DWELL_RATIO)
    period = schedule.cycle_length // 1
    dwell = schedule.hand_period * dwell_ratio
    intervals = Dict{Int,Vector{Tuple{Rational{Int},Rational{Int},Int}}}()
    for event in schedule.cycle_tosses
        event.kind == "empty" && continue
        event.ball === nothing && continue
        object_intervals = get!(intervals, event.ball, Tuple{Rational{Int},Rational{Int},Int}[])
        if event.kind == "hold"
            for (start, finish) in split_interval(event.beat // 1, event.height // 1, period)
                push!(object_intervals, (start, finish, event.from_body))
            end
        else
            for (start, finish) in split_interval(event.beat + event.height - dwell, dwell, period)
                push!(object_intervals, (start, finish, event.to_body))
            end
        end
    end
    objects = sort(collect(keys(intervals)))
    segments = [intervals[object] for object in objects]
    boundaries = Rational{Int}[0 // 1, period]
    for object_segments in segments, (start, finish, _body) in object_segments
        push!(boundaries, start, finish)
    end
    sort!(unique!(boundaries))
    body_ticks = [zeros(Rational{Int}, schedule.ball_count + 1) for _ in 1:schedule.body_count]
    for (start, finish) in zip(boundaries[1:end-1], boundaries[2:end])
        held_by_body = zeros(Int, schedule.body_count)
        for object_segments in segments
            match = findfirst(segment -> segment[1] <= start && finish <= segment[2], object_segments)
            if match !== nothing
                held_by_body[object_segments[match][3] + 1] += 1
            end
        end
        for body in 1:schedule.body_count
            body_ticks[body][held_by_body[body] + 1] += finish - start
        end
    end
    return [ticks ./ period for ticks in body_ticks]
end

File: pages/20260824170700_holding_topology_passing/analysis/ruby/passing_occupancy.rb
require_relative "passing_schedule"

module PassingAnalysis
  module Occupancy
    module_function

    def shares(schedule, dwell_ratio:)
      period = schedule.cycle_length
      dwell = schedule.hand_period * Rational(dwell_ratio)
      intervals = Hash.new { |table, object| table[object] = [] }
      schedule.cycle_tosses.each do |event|
        next if event.kind == "empty" || event.ball.nil?

        if event.kind == "hold"
          intervals[event.ball] << [event.beat, event.height]
        else
          intervals[event.ball] << [event.beat + event.height - dwell, dwell]
        end
      end
      segments = intervals.keys.sort.map do |object|
        intervals.fetch(object).flat_map { |start, length| split_interval(start, length, period) }
      end
      boundaries = ([0, period] + segments.flatten).uniq.sort
      occupancy_ticks = Array.new(schedule.ball_count + 1, 0.to_r)
      boundaries.each_cons(2) do |start, finish|
        held = segments.count do |object_segments|
          object_segments.any? { |held_start, held_finish| held_start <= start && finish <= held_finish }
        end
        occupancy_ticks[held] += finish - start
      end
      shares = occupancy_ticks.map { |ticks| ticks / period }
      {
        object_count: schedule.ball_count,
        occupancy_shares: shares,
        p_alpha: shares.first,
        p_polymorphy: shares[1...-1].sum(0.to_r),
        p_kappa: shares.last
      }
    end

    def body_shares(schedule, dwell_ratio:)
      period = schedule.cycle_length
      dwell = schedule.hand_period * Rational(dwell_ratio)
      intervals = Hash.new { |table, object| table[object] = [] }
      schedule.cycle_tosses.each do |event|
        next if event.kind == "empty" || event.ball.nil?

        if event.kind == "hold"
          intervals[event.ball] << [event.beat, event.height, event.from_body]
        else
          intervals[event.ball] << [event.beat + event.height - dwell, dwell, event.to_body]
        end
      end
      objects = intervals.keys.sort
      segments = objects.map do |object|
        intervals.fetch(object).flat_map do |start, length, body|
          split_interval(start, length, period).map { |held_start, held_finish| [held_start, held_finish, body] }
        end
      end
      boundaries = [0.to_r, period.to_r]
      segments.each do |object_segments|
        object_segments.each do |held_start, held_finish, _body|
          boundaries << held_start << held_finish
        end
      end
      boundaries = boundaries.uniq.sort
      body_ticks = Array.new(schedule.body_count) { Array.new(schedule.ball_count + 1, 0.to_r) }
      boundaries.each_cons(2) do |start, finish|
        held_by_body = Array.new(schedule.body_count, 0)
        segments.each do |object_segments|
          match = object_segments.find { |held_start, held_finish, _body| held_start <= start && finish <= held_finish }
          held_by_body[match[2]] += 1 if match
        end
        schedule.body_count.times do |body|
          body_ticks[body][held_by_body[body]] += finish - start
        end
      end
      body_ticks.map { |ticks| ticks.map { |count| count / period } }
    end

    def split_interval(start, length, period)
      return [] if length.zero?
      return [[0.to_r, period]] if length >= period

      normalized = start % period
      finish = normalized + length
      return [[normalized, finish]] if finish <= period

      [[normalized, period], [0.to_r, finish - period]]
    end
  end
end
